File: src/main/java/me/nasukhov/bot/command/LearnWord.java
package me.nasukhov.bot.command;

import me.nasukhov.bot.io.Input;
import me.nasukhov.bot.io.Output;
import me.nasukhov.study.ProgressRepository;
import me.nasukhov.dictionary.DictionaryRepository;
import me.nasukhov.dictionary.Word;

import java.util.ArrayList;
import java.util.List;

public class LearnWord implements Handler {
    private static final String NO_MORE_UNLEARNED_WORDS = "Вы изучили все слова из нашего словаря - больше новых слов нет.";

    private final DictionaryRepository dictionary;
    private final ProgressRepository progressRepository;

    public LearnWord(DictionaryRepository dictionaryRepository, ProgressRepository progressRepository) {
        this.dictionary = dictionaryRepository;
        this.progressRepository = progressRepository;
    }

    @Override
    public boolean supports(Input input) {
        return input.isDirectCommand("learn");
    }

    @Override
    public void handle(Input input, Output output) {
        if (!supports(input)) {
            return;
        }

        int lastLearnedWord = progressRepository.getLastLearnedWordId(input.channel());
        List<Integer> newWordsIds = new ArrayList<>();

        StringBuilder sb = new StringBuilder();
        for (Word word : dictionary.getChunk(3, lastLearnedWord)) {
            sb.append(word.word);
            sb.append(" - ");
            sb.append(word.translation);

            // TODO show description only for ambiguous words or words with high complexity(indicate in db?)
            //sb.append(word.description);
            sb.append("\n\n");
            newWordsIds.add(word.id);
        }

        if (sb.isEmpty()) {
            output.write(NO_MORE_UNLEARNED_WORDS);

            return;
        }

        progressRepository.setLastLearnedWords(input.channel(), newWordsIds);

        // Removing trailing newlines
        sb.delete(sb.length() - 2, sb.length());
        output.write(sb.toString());
    }
}

File: src/main/kotlin/me/nasukhov/tukitalearner/bot/LearnWord.kt
package me.nasukhov.tukitalearner.bot

import me.nasukhov.bot.command.Handler
import me.nasukhov.bot.io.Input
import me.nasukhov.bot.io.NoOutput
import me.nasukhov.bot.io.Output
import me.nasukhov.bot.io.TextOutput
import me.nasukhov.tukitalearner.dictionary.DictionaryRepository
import me.nasukhov.tukitalearner.study.Group
import me.nasukhov.tukitalearner.study.ProgressTracker
import org.springframework.data.domain.PageRequest
import org.springframework.stereotype.Component
import org.springframework.transaction.annotation.Transactional

@Component
class LearnWord(
    private val dictionary: DictionaryRepository,
    private val progressTracker: ProgressTracker,
) : Handler {
    companion object {
        private const val NO_MORE_UNLEARNED_WORDS = "Вы изучили все слова из нашего словаря - больше новых слов нет."
    }

    override fun supports(input: Input): Boolean = input.isDirectCommand("learn")

    @Transactional
    override fun handle(input: Input): Output {
        if (!supports(input)) {
            return NoOutput()
        }

        val group = Group(input.channel.id)
        val lastLearnedWord = progressTracker.getLastLearnedWordId(group)

        val sb = StringBuilder()
        val newWords = dictionary.findWords(lastLearnedWord, PageRequest.of(0, 3))
        for (word in newWords) {
            sb.append(word.word)
            sb.append(" - ")
            sb.append(word.translation)

            sb.append("\n\n")
        }

        if (sb.isEmpty()) {
            return TextOutput(NO_MORE_UNLEARNED_WORDS)
        }

        progressTracker.setLastLearnedWords(group, newWords)

        // Removing trailing newlines
        sb.delete(sb.length - 2, sb.length)

        return TextOutput(sb.toString())
    }
}
